src/NuGet.Clients/NuGet.SolutionRestoreManager/RestoreManagerPackage.cs

```diff
@@ -27,7 +28,7 @@ namespace NuGet.SolutionRestoreManager
     [PackageRegistration(UseManagedResourcesOnly = true, AllowsBackgroundLoading = true)]
     [ProvideAutoLoad(VSConstants.UICONTEXT.SolutionExists_string)]
     [Guid(PackageGuidString)]
-    public sealed class RestoreManagerPackage : AsyncPackage
+    public sealed class RestoreManagerPackage : AsyncPackage, IVsUpdateSolutionEvents4
```

Comment:
Why this is in the package? I'd suggest moving all the logic onto the worker. No need to split the responsibilities and define new contract.